L83-core-cpp-tls-key-export-feature.md

```diff
@@ -0,0 +1,93 @@
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+The proposal extends the existing `grpc::experimental::TlsCredentialsOptions` class with methods to set configuration for supporting Tls Key export at run-time. The proposed changes
+include:
+
+* A `grpc_tls_key_log_format` enum is added to `include/grpc/grpc_security_constants.h`
+  ```
+  /** Supported TLS Keylogging formats **/
+  typedef enum {
+    // Default key logging formal is NSS compatible.
+    TLS_KEY_LOG_FORMAT_NSS = 0,
+    // More formats can be added here for future extensions if desired.
+    TLS_KEY_LOG_FORMAT_NONE
+  } grpc_tls_key_log_format;
+  ```
+
+* A `grpc::experimental::TlsKeyLoggerConfig` struct is defined in `include/grpcpp/security/tls_credentials_options.h` to support configuration for Key export. It currently has two members:
+  * `std::string tls_key_log_file_path`: Full path where the TLS keys should be exported to.
+  * `grpc_tls_key_log_formattls_key_log_format`: Format of the exported keys.
+
+* Adding a `set_tls_key_log_config(grpc::experimental::TlsKeyLoggerConfig cfg)` method to the `grpc::experimental::TlsCredentialsOptions` class to set the desired configuration on the client/server side.
+
+These changes allow specifying TLS Key export configuration on a per-channel level on the client side and on a per-port level on the server side. If the configuration is not explicitly set, TLS key logging is disabled. This config  struct can be extended in the future to support advanced features such as key logging based on source ip filtering etc.
+
+## Example Usage
+
+# Client Side (per-channel)
+
+```
+grpc::experimental::TlsChannelCredentialsOptions options;
+grpc::experimental::TlsKeyLoggerConfig config;
+config.set_tls_key_log_file_path(
+  absl::StrCat(BASE_DIR, "/ssl-key-log-client.txt"));
+config.set_tls_key_log_format(TLS_KEY_LOG_FORMAT_NSS);
+options.set_tls_key_log_config(config);
+auto channel = grpc::CreateChannel(server_address,
+  grpc::experimental::TlsCredentials(options)));
+```
+
+# Server Side (per-port)
+
+```
+grpc::experimental::TlsServerCredentialsOptions options(certificate_provider);
+grpc::experimental::TlsKeyLoggerConfig config;
+config.set_tls_key_log_file_path(
+  absl::StrCat(BASE_DIR, "/ssl-key-log-server.txt"));
+config.set_tls_key_log_format(TLS_KEY_LOG_FORMAT_NSS);
+options.set_tls_key_log_config(config);
+...
+ServerBuilder builder;
+builder.AddListeningPort(server_address,
+      grpc::experimental::TlsCredentials(options));
+```
+
+
+## Rationale
+
+Alternatives using ENVIRONMENT Variables to set TLS Key export configurations were considered. These allow enabling/disabling key export without having to rebuild the binary. However the disadvantages here include:
```

Comment:
I'm still an advocate for requiring both a channel arg and an environment variable... having two locks on such a potentially damaging feature seems useful.

Reply:
You mean like an environment variable override i.e, log keys only when both the env variable and the above configs are set ?

Reply:
Craig, I agree it is a potentially damaging feature if misused. But in the current design, in order to enable it, a user needs to explicitly create a `TlsKeyLoggerConfig` instance and set it in `TlsServerCredentialsOptions`, which I think is sufficient to confirm user's intention to enable the feature. I am concerned that introducing extra knobs may result in usability problems.

Reply:
There is a GRPC_TLS_KEY_LOGGING_ENABLED environment variable to override right now. I think we can remove this at a later point if there are usability issues.